add-friends-likes

puffpundb opened the pull request:
Создан сервисный слой.
Контроллеры доведены до соответствия с REST.

puffpundb committed:
Переработана архитектора в соответствии с ТЗ.
Создан сервисный слой.
Контроллеры доведены до соответствия с REST.

puffpundb committed:
Поправлен кодстайл

puffpundb committed:
Поправлен кодстайл #2

puffpundb committed:
Поправлен кодстайл #3

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,115 +1,65 @@
     @GetMapping
-    public ArrayList<Film> getAllFilm() {
-        return new ArrayList<>(filmStore.values());
+    @ResponseStatus(HttpStatus.OK)
```
Статус OK по умолчанию для успешных запросов

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,115 +1,65 @@
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void dislikedFilm(@PathVariable Integer id, @PathVariable Integer userId) {
+        if (!userStore.isContain(userId)) {
+            throw new NotFoundException("Пользователь не найден");
```
Это логика сервиса

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,115 +1,65 @@
-    private final Map<Integer, Film> filmStore = new HashMap<>();
+    private final FilmStorage filmStore;
+    private final FilmService filmService;
+    private final UserStorage userStore;
```
В контроллерах обращаемся к сервисам, в сервисах - к стораджам, соблюдаем цепочку взаимодействия controller -> service -> storage

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,112 +1,61 @@
-    private final Map<Integer, User> userDataBase = new HashMap<>();
-    private Integer currentMaxId = 0;
+    private final UserStorage userStore;
+    private final UserService userService;
```
Аналогично, работаем с сервисами

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -15,4 +17,11 @@ public class Film {
+
+    private Set<Integer> usersLike = new HashSet<>();
+
+    @Override
+    public int compareTo(Film o) {
+        return Integer.compare(o.getUsersLike().size(), this.usersLike.size());
+    }
```
Лучше ввести счетчик лайков, инкапсулировать логику добавления/удаления лайков в `usersLike` и обновлять счетчик по длине `usersLike`

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmServiceImpl.java:
```diff
@@ -0,0 +1,43 @@
+@Service
+@AllArgsConstructor
+public class FilmServiceImpl implements FilmService {
+	FilmStorage filmDataBase;
```
Модификаторы доступа

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmServiceImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+	@Override
+	public void liked(Integer filmId, Integer userId) {
+		if (filmDataBase.isContain(filmId)) {
+			filmDataBase.getFilm(filmId).getUsersLike().add(userId);
+			return;
+		}
+
+		throw new NotFoundException("Фильм не найден");
```
во-первых, на уровне сервиса нужно проверить пользователя, во-вторых, метод `getFilm` можно реализовать таким образом, чтобы он возвращал объект, если он найден в хешмапе или выбрасывал исключение - если нет. 
А вообще, хорошая практика воспользоваться `Optional`, это очень удобный класс для работы с nullable объектами

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmServiceImpl.java:
```diff
@@ -0,0 +1,43 @@
+	public List<Film> getMostRatedFilms(int count) {
+		return filmDataBase.getAllFilm().stream()
+				.sorted()
+				.limit(count)
```
Стоит проверить на неотрицательность, иначе упадем на исключении

puffpundb committed:
Перенесена логика сервисов в соответствующие сервисы.
Создана правильная цепочка взаимодействий сервисов, контроллеров и хранилищ.
Минорные исправления логики.

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmServiceImpl.java:
```diff
@@ -3,28 +3,34 @@
 
 	@Override
 	public void disliked(Integer filmId, Integer userId) {
+		if (!userDataBase.isContain(userId)) {
+			throw new NotFoundException("Пользователь с данным id не найден");
+		}
```
Может стоит сделать реализацию метода getUser() аналогично getFilm() с проверкой и выбросом исключения?

Или можно просто проверить на null полученного юзера, тогда метод `isContain` вообще не нужен

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmServiceImpl.java:
```diff
@@ -3,28 +3,34 @@
+			throw new NotFoundException("Пользователь с данным id не найден");
+		}
+
 		if (filmDataBase.isContain(filmId)) {
```
Эта проверка и выброс исключения уже не нужны, в метод getFilm это входит

puffpundb committed:
Минорные изменения:
Исправлена реализация getUser() по аналогии с getFilm().
Убраны лишние исключения.
